File: DataStructures/stack.py
import random
import numpy as np

# Stack
class Stack:

    # Initializes size of stack and allocates memory
    def __init__(self, max=10) -> None:
        self.max = max
        self.top = -1
        self.stack = [None for x in range(max)]

    # Is stack full?
    def isOverflow(self) -> bool:
        if self.top >= self.max-1:
            print("Stack Overflow")
            return True
        return False

    # Is stack empty?
    def isUnderflow(self) -> bool:
        if self.top <= -1:
            print("Stack Underflow")
            return True
        return False

    # Pushes data on top of stack
    def push(self, data) -> None:
        if not self.isOverflow():
            self.top += 1
            self.stack[self.top] = data

    # Return data if stack is not empty
    def pop(self):
        if not self.isUnderflow():
            data = self.stack[self.top]
            self.stack[self.top] = None
            self.top -= 1
            return data
        return None

    # Returns top most element
    def displayTop(self):
        if not self.isUnderflow():
            return self.stack[self.top]
        return None

    # Returns stack elements
    def getStack(self) -> list:
        return [i for i in self.stack if i is not None]

    # Displays stack
    def display(self) -> None:
        if self.isUnderflow():
            print("Stack is Empty")
            return
        stack = self.getStack()[::-1]
        print(stack[0], "<- top")
        for data in stack[1:]:
            print(data)

# Main
if __name__ == "__main__":
    GREEN = '\033[92m'
    RESET = '\033[0m'
    print(f"{GREEN}Stack{RESET}")

    stack = Stack(5)

    print("Size of Stack is 5")
    
    # Generating Random Data
    data = [random.randint(0,100) for x in range(0,6)]

    # Displaying Random Data Generated
    print(f"Random Data Generated : {data}")

    # Pushing Data
    print("\nPushing Data ...")
    for i in data:
        stack.push(i)

    # Displaying Stack
    print("\nStack: ")
    stack.display()

    # Popping Data from Stack
    print("\nPopping data: ")
    for j in range(0,6):
        poppedData = stack.pop()

        if poppedData is None:
            break

        print(f"Data Popped : {poppedData}")

File: DataStructures/stack.go
package main

import (
	"fmt"
	"log"
)

type Stack struct {
	items []string
	top   int `default:"-1"`
	MAX   int
}

func CreateStack(size int) *Stack {
	return &Stack{
		items: make([]string, size+1),
		MAX:   size,
	}
}

func (s *Stack) IsOverflow() bool {
	if s.top < s.MAX {
		return false
	}

	log.Println("Stack overflow")
	return true
}

func (s *Stack) IsUnderflow() bool {
	if s.top > -1 {
		return false
	}

	log.Println("Stack underflow")
	return true
}

func (s *Stack) Push(val string) bool {
	if s.IsOverflow() {
		return false
	}

	s.top++
	s.items[s.top] = val

	return true
}

func (s *Stack) Pop() string {
	if s.IsUnderflow() {
		return ""
	}

	s.top--
	return s.items[s.top+1]
}

func (s *Stack) Print() {
	fmt.Print("Stack => ")
	for i := 0; i <= s.top; i++ {
		fmt.Printf("%s ", s.items[i])
	}
	fmt.Println()
}

func (s *Stack) Top() string {
	if s.IsUnderflow() {
		return ""
	}

	return s.items[s.top]
}

func main() {
	s := CreateStack(5)

	s.Push("10")
	s.Push("20")
	s.Push("30")
	s.Push("40")
	s.Push("50")

	fmt.Printf("Top => %s\n", s.Top())

	s.Print()

	fmt.Printf("Popped => %s\n", s.Pop())
	fmt.Printf("Top => %s\n", s.Top())

	fmt.Printf("Popped => %s\n", s.Pop())

	s.Print()
}
